docs/data/data-grid/performance/performance.md

```diff
@@ -2,43 +2,44 @@
 
 <p class="description">Improve the performance of the DataGrid using the recommendations from this guide.</p>
 
-## Memoize inner components with `React.memo`
+## Extract static objects and memoize root props
 
-The `DataGrid` component is composed of a central state object where all data is stored.
-When an API method is called, a prop changes, or the user interacts with the UI (e.g. filtering a column), this state object is updated with the changes made.
-To reflect the changes in the interface, the component must re-render.
-Since the state behaves like `React.useState`, the `DataGrid` component will re-render its children, including column headers, rows, and cells.
-With smaller datasets, this is not a problem for concern, but it can become a bottleneck if the number of rows increases, especially if many columns render [custom content](/x/react-data-grid/column-definition/#rendering-cells).
-One way to overcome this issue is using `React.memo` to only re-render the child components when their props have changed.
-To start using memoization, import the inner components, then pass their memoized version to the respective slots, as follow:
+The `DataGrid` component uses `React.memo` to optimize its performance, which means itself and its subcomponents only
+re-render when their props change. But it's very easy to cause unnecessary re-renders if the root props of your
+`DataGrid` aren't memoized. Take the example below, the `slots` and `initialState` objects are re-created on every
+render, which means the `DataGrid` itself has no choice but to re-render as well.
 
 ```tsx
-import {
-  GridRow,
-  GridColumnHeaders,
-  DataGrid, // or DataGridPro, DataGridPremium
-} from '@mui/x-data-grid';
-
-const MemoizedRow = React.memo(GridRow);
-const MemoizedColumnHeaders = React.memo(GridColumnHeaders);
-
-<DataGrid
-  slots={{
-    row: MemoizedRow,
-    columnHeaders: MemoizedColumnHeaders,
-  }}
-/>;
+function Component({ rows }) {
+  return (
+    <DataGrid
+      rows={rows}
+      slots={{
+        row: CustomRow,
+      }}
+      cellModesModel={{ [rows[0].id]: { name: { mode: GridCellModes.Edit } } }}
+    />
+  );
+}
 ```
 
-The following demo show this trick in action.
-It also contains additional logic to highlight the components when they re-render.
+An easy way to prevent re-renders is to extract any object that can be a static object, and to memoize any object that
+depends on another object. This applies to any prop that is an object or a function.
 
-{{"demo": "GridWithReactMemo.js", "bg": "inline", "defaultCodeOpen": false}}
```

Comment:
Would it make sense to keep this demo to visualize the rerenders?

Reply:
It's not accurate anymore. If we want to visualize re-renders, we'd need to add the tracking code inside the rows/cells. Thanks to the new architecture, the cells can re-render even if their parent hasn't re-rendered. But I think I could implement it by adding a `useEffect` inside `.renderCell`.

Reply:
Maybe this would work?
```jsx
const CellWithTracer = React.forwardRef((props, ref) => {
  return <TraceUpdates ref={ref} Component={GridCell} {...props} />;
});
const MemoizedCell = React.memo(CellWithTracer);
const slots = { cell: MemoizedCell }
```

Reply:
Not really, some of the cell updates now come from within the cell hooks, not because their parent re-rendered & changed their props.

Reply:
I've re-added the visualization here: https://deploy-preview-9037--material-ui-x.netlify.app/x/react-data-grid/performance/#visualization

It's cell-based now, so we can clearly see that only the required cells re-render.

Reply:
> Not really, some of the cell updates now come from within the cell hooks, not because their parent re-rendered & changed their props.

Oh, right.

> I've re-added the visualization here: [deploy-preview-9037--material-ui-x.netlify.app/x/react-data-grid/performance/#visualization](https://deploy-preview-9037--material-ui-x.netlify.app/x/react-data-grid/performance/#visualization)

> It's cell-based now, so we can clearly see that only the required cells re-render on scroll, both vertically and horizontally.

Looks awesome! 🎉
How did you make it work for the updates coming from the cell hooks?

Reply:
`useGridSelector` now encapsulates a `useState`, and the store thing calls the `setState` function when the selector value changes, so the cells can update only when their `useGridSelector`ed value changes.

Reply:
Yep, works thanks to the wrapper. For the tracing to work we need custom code inside the component, there is no way around that.